File: python_ruby/Class_Member/3.py
class Calc(object):
    _history = []
    def __init__(self,value1,value2):
        if isinstance(value1, int):
            self.value1 = value1
        if isinstance(value2, int):
            self.value2 = value2
    def add(self):
        result = self.value1 + self.value2
        Calc._history.append("add : %d + %d = %d" % (self.value1,self.value2,result))
        return result
    def subtract(self):
        result = self.value1 - self.value2
        Calc._history.append("subtract : %d - %d = %d" % (self.value1,self.value2,result))
        return result
    def setValue1(self,value1):
        if isinstance(value1,int):
            self.value1 = value1
    def gatValue1(self):
        return self.value1
    @classmethod
    def history(cls):
        for item in Calc._history:
            print(item)
class CalcMultiply(Calc):
    def multiply(self):
        result = self.value1 * self.value2
        Calc._history.append("multiply : %d * %d = %d" % (self.value1,self.value2,result))
        return result
class CalcDivide(CalcMultiply):
    def divide(self):
        result = self.value1 / self.value2
        Calc._history.append("divide : %d / %d = %d" % (self.value1,self.value2,result))
        return result

calc = CalcMultiply(10,10)
print(calc.add())
print(calc.subtract())
print(calc.multiply())
calc1 = CalcDivide(20,10)
print(calc1.divide())
Calc.history()

File: python_ruby/Class_Member/3.rb
class Calc
  attr_reader :value1, :value2
  attr_writer :value1
  @@_history = []
  def initialize(value1,value2)
    @value1 = value1
    @value2 = value2
  end

  def add()
    result = @value1 + @value2
    @@_history.push("add : #{@value1} + #{@value2} = #{result}")
    return result
  end

  def subtract()
    result = @value1 - @value2
    @@_history.push("subtract : #{@value1} - #{@value2} = #{result}")
    return result
  end

  def setValue1(value1)
    if value1.is_a?(Integer)
      @value1 = value1
    end
  end

  def getValue1()
    return @value1
  end

  def Calc.history()
    for item in @@_history
      p item
    end

  end
end

class CalcMultiply < Calc
  def multiply()
    result = @value1 * @value2
    @@_history.push("multiply : #{@value1} * #{@value2} = #{result}")
    return result
  end
end

class CalcDivide < CalcMultiply
  def divide()
    result = @value1 / @value2
    @@_history.push("divide : #{@value1} / #{@value2} = #{result}")
    return result
  end
end

calc = CalcMultiply.new(10,10)
p calc.add()
p calc.multiply()

calc1 = CalcDivide.new(20,20)
p calc1.add()
p calc1.multiply()
p calc1.divide()
calc1.setValue1(40)
p calc1.divide()
p calc1.getValue1()
Calc.history()
